src/StoreApi/Utilities/NotPurchasableException.php

```diff
@@ -0,0 +1,73 @@
+<?php
+namespace Automattic\WooCommerce\Blocks\StoreApi\Utilities;
+
+/**
+ * NotPurchasableException class.
+ *
+ * @internal This API is used internally by Blocks, this exception is thrown when an item in the cart is not able to be
+ * purchased.
+ */
+class NotPurchasableException extends \Exception {
```

Comment:
The reason the exceptions are separate like this is because we need to catch them separately in `validate_cart_items` so we can generate the correct error messages. I agree that we should create a new exception to reduce code duplication in these exceptions, but I don't think we can catch the parent exception for the reason stated above.

I propose we rename what is currently `StockAvailabilityException` which is the **ultimate** exception, it contains the `WP_Error` object with all of the error messages inside, to something like `InvalidQuantitiesInCartException`. Then we can use the name `StockAvailabilityException` as the parent of the `OutOfStockException`, `PartialOutOfStockException`, `NotPurchasableException`, and `TooManyInCartException` exceptions.